---
title: "3D Breast Cancer Spatial Transcriptomics Analysis with MERingue"
author: "Jean Fan"
date: "`r Sys.Date()`"
# rmarkdown::render('vignettes/BCL_analysis.Rmd', clean=FALSE, output_format='all')
output:
  html_document:
    keep_md: yes
  pdf_document: 
    highlight: zenburn
  md_document:
    variant: markdown_github
vignette: >
  %\VignetteIndexEntry{3D Breast Cancer Spatial Transcriptomics Analysis with MERingue}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, options, include = FALSE}
library(knitr)
opts_chunk$set(
    fig.path = 'figure/',
    cache.path = 'cache/',
    cache = TRUE, 
    results = "hold"
)
```

In this vignette, we will walk through an analysis of spatial transcriptomics data for 4 consecutive tissue slices of a breast cancer biopsy. Briefly, for spatial transcriptomics, histological sections are placed on a grid of poly(dT) probe spots approximately 100 um in diameter, each with a unique DNA barcode. By resolving the DNA barcodes, spatial transcriptomics enables matching of detected mRNA abundances with their original spatially resolved spot, resulting in full transcriptome RNA-sequencing data with homogenously-spaced two-dimensional positional information at the pixel level. See the [original publication](http://science.sciencemag.org/content/353/6294/78) for more information.

The data has been prepared for you and is available as a part of the package. Here, each dataset is a list containing two items: `pos` is a dataframe where each row is a probe spot's x and y positions in space, and `counts` is a counts matrix where each column is a probe spot and each row is a gene.

```{r, bcl-data}
library(MERingue)
data(BCL1) # layer 1
data(BCL2) # layer 2
data(BCL3) # layer 3
data(BCL4) # layer 4
```

We will first combine the data from all four samples into a single counts matrix. We will also combine the position information into a list of positions. 

```{r, bcl-combine}
# Get common set of genes
genes.have <- Reduce(intersect, list(
  rownames(BCL1$counts),
  rownames(BCL2$counts),
  rownames(BCL3$counts),
  rownames(BCL4$counts)

))
# Combine into large counts matrix
counts <- cbind(
  BCL1$counts[genes.have,],
  BCL2$counts[genes.have,],
  BCL3$counts[genes.have,],
  BCL4$counts[genes.have,]
)
# Unique names based on layers
colnames(counts) <- c(
  paste0('L1-', colnames(BCL1$counts)),
  paste0('L2-', colnames(BCL2$counts)),
  paste0('L3-', colnames(BCL3$counts)),
  paste0('L4-', colnames(BCL4$counts))
)
# Layer factor
layer <-  c(
  rep('L1', ncol(BCL1$counts)),
  rep('L2', ncol(BCL2$counts)),
  rep('L3', ncol(BCL3$counts)),
  rep('L4', ncol(BCL4$counts))
)
names(layer) <- colnames(counts)
# List of positions
posList <- list(
  BCL1$pos[colnames(BCL1$counts),],
  BCL2$pos[colnames(BCL2$counts),],
  BCL3$pos[colnames(BCL3$counts),],
  BCL4$pos[colnames(BCL4$counts),]
)
rownames(posList[[1]]) <- paste0('L1-', rownames(posList[[1]]))
rownames(posList[[2]]) <- paste0('L2-', rownames(posList[[2]]))
rownames(posList[[3]]) <- paste0('L3-', rownames(posList[[3]]))
rownames(posList[[4]]) <- paste0('L4-', rownames(posList[[4]]))
```

We will then filter out genes with fewer than 10 counts across all probe spots and filter out probe spots with fewer than 1500 total counts and update our position information accordingly. 

```{r, bcl-clean, fig.width=8, fig.height=3}
# Very stringent filter here
cc <- cleanCounts(counts, min.reads = 10, min.lib.size = 1500, plot=TRUE)
mat <- normalizeCounts(cc, log=FALSE, verbose=TRUE)
posList[[1]] <- posList[[1]][intersect(rownames(posList[[1]]), colnames(mat)),]
posList[[2]] <- posList[[2]][intersect(rownames(posList[[2]]), colnames(mat)),]
posList[[3]] <- posList[[3]][intersect(rownames(posList[[3]]), colnames(mat)),]
posList[[4]] <- posList[[4]][intersect(rownames(posList[[4]]), colnames(mat)),]
```

Now, let's visualize the probe spots for each layer. For demonstrative purposes, we will look for evidence of significant spatial aggregation for 1000 randomly chosen genes. 

```{r, bcl-init, fig.width=8, fig.height=8}
# Plot
par(mfrow=c(2,2), mar=rep(2,4))
plotEmbedding(posList[[1]], groups=layer, main='Layer 1', cex=2)
plotEmbedding(posList[[2]], groups=layer, main='Layer 2', cex=2)
plotEmbedding(posList[[3]], groups=layer, main='Layer 3', cex=2)
plotEmbedding(posList[[4]], groups=layer, main='Layer 4', cex=2)

# Sample 1000 genes for demonstrative purposes only
set.seed(0)
test <- sample(rownames(mat), 1000)
mat <- mat[test,]
```

Now, we can analyze each layer separately to look for significantly spatially aggregated genes affecting more than 5% of cells. For each layer, we will construct a binary weight matrix that represents the spatial relationship

```{r, bcl-indLayer, fig.width=8, fig.height=8}
helper <- function(pos, mat) {
  w <- voronoiAdjacency(pos, filterDist = 1.5, plot=TRUE)
  # get spatially clustered genes
  I <- getSpatialPatterns(mat, w)
  # filter for significant hits driven by more than 10% of cells
  results.filter <- filterSpatialPatterns(mat = mat,
                                          I = I,
                                          w = w,
                                          adjustPv = TRUE,
                                          alpha = 0.05,
                                          minPercentCells = 0.05,
                                          verbose = TRUE)
  # return results
  list(I=I, sig.genes=results.filter)
}

# Analyze each layer using helper function
par(mfrow=c(2,2), mar=rep(2,4))
L1 <- helper(posList[[1]], mat[, rownames(posList[[1]])])
L2 <- helper(posList[[2]], mat[, rownames(posList[[2]])])
L3 <- helper(posList[[3]], mat[, rownames(posList[[3]])])
L4 <- helper(posList[[4]], mat[, rownames(posList[[4]])])
````

We can also look specifically for patterns that are consistent across layers.

```{r, bcl-crosslayer, fig.width=8, fig.height=8}
cw <- getCrossLayerNeighbors(posList, k=3)

# Visualize 3D network
# Note requires rgl library
plot3dNetwork <- function(posList, w) {
  library(rgl)
  pos3d <- Reduce(rbind, posList)
  z <- unlist(lapply(seq_along(posList), function(i) { rep(i, nrow(posList[[i]])) }))
  pos3d <- cbind(pos3d, z)
  head(pos3d)
  x <- pos3d[,1]
  y <- pos3d[,2]
  z <- pos3d[,3]
  names(x) <- names(y) <- names(z) <- rownames(pos3d)
  rgl::plot3d(x, y, z, col=rainbow(4)[as.factor(pos3d[,3])], size=5, alpha = 0.50,
         xlab='', ylab='', zlab='')
  idx <- which(w>0, arr.ind = T)
  for(i in seq_len(nrow(idx))) {
    n1 <- rownames(w)[idx[i,1]]
    n2 <- rownames(w)[idx[i,2]]
    rgl::segments3d(
      x=c(x[n1], x[n2]),
      y=c(y[n1], y[n2]),
      z=c(z[n1], z[n2]),
      col='grey'
    )
  }
}
# Uncomment to plot
# plot3dNetwork(posList, cw)

# Look for genes exhibiting spatial autocorrelation across layers
I <- getSpatialPatterns(mat, cw)
results.filter <- filterSpatialPatterns(mat = mat,
                                        I = I,
                                        w = cw,
                                        adjustPv = TRUE,
                                        alpha = 0.05,
                                        minPercentCells = 0.1/4,
                                        verbose = TRUE)
cross <- list(I=I, sig.genes=results.filter)
```

We can now compare our results from our individual sample analyses and our cross sample analyses using an Upset plot. Upset plots are useful for visualizing complex set intersections, much like a venn diagram. Here, looking at the last column, we can see 2 genes that are significantly spatially clustered in all layers AND also consistent across layers. We can also see 1 gene that is significantly spatially clustered in layer 1, 3, and 4, but not in layer 2 and not consistent across layers. 

```{r, bcl-compare}
## compare
library(UpSetR)
sig <- list('L1'=L1$sig.genes, 
            'L2'=L2$sig.genes, 
            'L3'=L3$sig.genes, 
            'L4'=L4$sig.genes, 
            'Cross'=cross$sig.genes)
upset(fromList(sig))
```

Let's look at a gene that is significant within all layers AND consistent across layers compared to a gene that is significant within multiple layers but NOT consistent across layers.

```{r, bcl-sample, fig.width=8, fig.height=8}
gall <- Reduce(intersect, 
               list('L1'=L1$sig.genes, 
                    'L2'=L2$sig.genes, 
                    'L3'=L3$sig.genes, 
                    'L4'=L4$sig.genes))
gdups <- unlist(list('L1'=L1$sig.genes, 
                     'L2'=L2$sig.genes, 
                     'L3'=L3$sig.genes, 
                     'L4'=L4$sig.genes))
gdups <- gdups[duplicated(gdups)]
gcross <- cross$sig.genes
g1 <- intersect(gcross, gall)[1]
print(g1)
g2 <- setdiff(gdups, gcross)[1]
print(g2)
```

```{r, bcl-consistent, fig.width=8, fig.height=8}
# Plot consistent gene
L1$I[g1,]
L2$I[g1,]
L3$I[g1,]
L4$I[g1,]
cross$I[g1,]
par(mfrow=c(2,2), mar=rep(2,4))
plotEmbedding(posList[[1]], 
              colors=scale(mat[g1, rownames(posList[[1]])])[,1], 
              main=paste0(g1, '\n Layer 1'), cex=2)
plotEmbedding(posList[[2]], 
              colors=scale(mat[g1, rownames(posList[[2]])])[,1], 
              main=paste0(g1, '\n Layer 2'), cex=2)
plotEmbedding(posList[[3]], 
              colors=scale(mat[g1, rownames(posList[[3]])])[,1], 
              main=paste0(g1, '\n Layer 3'), cex=2)
plotEmbedding(posList[[4]], 
              colors=scale(mat[g1, rownames(posList[[4]])])[,1], 
              main=paste0(g1, '\n Layer 4'), cex=2)
```

```{r, bcl-inconsistent, fig.width=8, fig.height=8}
# Plot inconsistent gene
L1$I[g2,]
L2$I[g2,]
L3$I[g2,]
L4$I[g2,]
cross$I[g2,]
par(mfrow=c(2,2), mar=rep(2,4))
plotEmbedding(posList[[1]], 
              colors=scale(mat[g2, rownames(posList[[1]])])[,1], 
              main=paste0(g2, '\n Layer 1'), cex=2)
plotEmbedding(posList[[2]], 
              colors=scale(mat[g2, rownames(posList[[2]])])[,1], 
              main=paste0(g2, '\n Layer 2'), cex=2)
plotEmbedding(posList[[3]], 
              colors=scale(mat[g2, rownames(posList[[3]])])[,1], 
              main=paste0(g2, '\n Layer 3'), cex=2)
plotEmbedding(posList[[4]], 
              colors=scale(mat[g2, rownames(posList[[4]])])[,1], 
              main=paste0(g2, '\n Layer 4'), cex=2)
```
