---
title: "mOB Analysis"
author: "Jean Fan"
date: "`r Sys.Date()`"
output:
  md_document:
    variant: markdown_github
vignette: >
  %\VignetteIndexEntry{mOB Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(knitr)
opts_chunk$set(
    fig.path = 'figure/',
    cache.path = 'cache/',
    cache = TRUE, 
    results = "hold"
)
```

In this vignette, we will walk through an analysis of spatial transcriptomics data for the mouse olfactory bulb. The data has been prepared for you and is available as a part of the package.

```{r, 'mOB_data', fig.width=6, fig.height=3}
library(MERingue)
library(Matrix)

## load mOB data
data(cd)
data(pos)

cd <- cd[,rownames(pos)]
counts <- cleanCounts(cd, min.reads=100)
mat <- normalizeCounts(counts)

par(mfrow=c(1,2), mar=rep(2,4))
plotEmbedding(pos, colors=colSums(mat), cex=1, main='Library Size')
plotEmbedding(pos, colors=colSums(mat>0), cex=1, main='Library Complexity')
```

We will define voxels as neighbors if they are mutual nearest neighbors in space with k=6.

```{r, 'mOB_neighbor', fig.width=5, fig.height=5}
## get neighbor relationship
w <- getSpatialWeights(pos, k=6)
par(mfrow=c(1,1), mar=rep(2,4))
plotNetwork(pos, w, main='MNN Network')
````

We can calculate the spatial autocorrelation for all genes.

```{r, 'mOB_autocorr', fig.width=6, fig.height=3}
## calculate spatial autocorrelation
start <- Sys.time()
I <- getSpatialPatterns(mat, w, verbose=TRUE)
end <- Sys.time()
message(paste0(nrow(mat), ' genes analyzed in ', difftime(end, start, units='mins'), ' minutes.'))
print(head(I))
```

We want to focus on significantly autocorrelated (adjusted p-value < 0.05) driven by more than 5% of cells. 

```{r, 'mOB_filter'}
## filter for patterns driven by more than 5% of cels
results.filter <- filterSpatialPatterns(mat = mat,
                                        I = I,
                                        adjustPv = TRUE,
                                        alpha = 0.05,
                                        minPercentCells = 0.05,
                                        verbose = TRUE)
print(head(I[results.filter,]))
```

We can visualize each gene individually. 

```{r, 'mOB_sample', fig.width=6, fig.height=3}
I.final <- I[results.filter,]
I.final <- I.final[order(I.final$p.value, decreasing=FALSE),]
par(mfrow=c(1,2), mar=rep(2,4))
invisible(interpolate(pos, mat[rownames(I.final)[1],], main=rownames(I.final)[1]))
````

Or group them into prominant patterns based on spatial cross-correlation.

```{r, 'mOB_crosscor'}
## spatial cross correlation to cluster genes
scc <- spatialCrossCorMatrix(as.matrix(mat[results.filter,]), w)
d <- as.dist((-scc - min(-scc))) ## convert to distance
```

```{r, 'mOB_group', fig.width=6, fig.height=18}
ggroup <- groupSigSpatialPatterns(pos=pos, 
                                  mat=as.matrix(mat[results.filter, ]), 
                                  d=d,
                                  minClusterSize=5,
                                  plot=TRUE)
```


